cli/src/torii/routing.rs

```diff
@@ -420,36 +420,57 @@ fn handle_metrics(sumeragi: &SumeragiHandle) -> Result<String> {
         .map_err(Error::Prometheus)
 }
 
-#[cfg(feature = "telemetry")]
-#[allow(clippy::unnecessary_wraps)]
-fn handle_status(sumeragi: &SumeragiHandle) -> Result<warp::reply::Json, Infallible> {
+fn update_metrics_gracefully(sumeragi: &SumeragiHandle) {
     if let Err(error) = sumeragi.update_metrics() {
         iroha_logger::error!(%error, "Error while calling `sumeragi::update_metrics`.");
     }
-    let status = Status::from(&sumeragi.metrics());
-    Ok(reply::json(&status))
 }
 
 #[cfg(feature = "telemetry")]
-#[allow(clippy::unused_async)]
-async fn handle_status_precise(sumeragi: SumeragiHandle, segment: String) -> Result<Json> {
-    if let Err(error) = sumeragi.update_metrics() {
-        iroha_logger::error!(%error, "Error while calling `sumeragi::update_metrics`.");
-    }
-    // TODO: This probably can be optimised to elide the full
-    // structure. Ideally there should remain a list of fields and
-    // field aliases somewhere in `serde` macro output, which can
-    // elide the creation of the value, and directly read the value
-    // behind the mutex.
+#[allow(clippy::unnecessary_wraps)]
+fn handle_status(
+    sumeragi: &SumeragiHandle,
+    accept: Option<impl AsRef<str>>,
+    subpath: &warp::path::Tail,
+) -> Result<reply::Response> {
+    use eyre::ContextCompat;
+    const PARITY_SCALE_MIME: &'_ str = "application/x-parity-scale";
+
+    update_metrics_gracefully(sumeragi);
     let status = Status::from(&sumeragi.metrics());
-    match serde_json::to_value(status) {
-        Ok(value) => Ok(value
-            .get(segment)
-            .map_or_else(|| reply::json(&value), reply::json)),
-        Err(err) => {
-            iroha_logger::error!(%err, "Error while converting to JSON value");
-            Ok(reply::json(&None::<String>))
+
+    let subpath = subpath.as_str();
+    if subpath.is_empty() {
+        if accept.is_some_and(|x| x.as_ref() == PARITY_SCALE_MIME) {
+            let body: hyper::Body = status.encode().into();
+
+            warp::http::Response::builder()
+                .status(StatusCode::OK)
+                .header(warp::http::header::CONTENT_TYPE, PARITY_SCALE_MIME)
+                .body(body)
+                .wrap_err("Failed to build response body")
+                .map_err(Error::StatusFailure)
+        } else {
+            Ok(reply::json(&status).into_response())
```

Comment:
I don't fully get you.

> Is it fine to not handle the sub-path here?

But this branch is for the case when the sub-path is empty, therefore nothing to handle.

> as specifying `Accept: application/json` will suddenly stop the feature from working

What feature will stop to work if request contains `Accept: application/json`?

I think it will help me if you provide an example code snippet.

So far, the only alternative I see to reduce if-else branching - unify handling of subpath. `"".split('/')` will result in an empty iterator, thus the full status will be returned anyway:

```rust
// *using unstable let expression*
if let Some(accept) = accept &&
    accept == PARITY_SCALE_MIME &&
    subpath.is_empty() {
    // return full status as scale
} else {
    let value = serde_json::to_value(status)
        .wrap_err("Failed to serialize JSON")
        .map_err(Error::StatusFailure)?;

    let reply = subpath
        .split('/')
        .try_fold(&value, serde_json::Value::get)
        .wrap_err_with(|| eyre!("Path not found: \"{}\"", subpath))
        .map(|segment| reply::json(segment).into_response())
        .map_err(Error::StatusSegmentNotFound)?;

    Ok(reply)
}
```

Less branching, but a little runtime overhead, which my be avoided by a branch, again:

```rust
if subpath.is_empty() { return Ok(reply::json(status)) }
```

Reply:
I think I had a brainfart there, mixing up the levels of nesting. Now when I look at it, this logic looks absolutely fine.

But I have another suggestion: return HTTP [406](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/406) error when a subpath and scale codec is requested, as returning json when scale is requested may be kinda unexpected.

Reply:
Yeah, 406 fits, and Iroha don't really need to be "gentle" with default fallbacks.

However, I just realized that there are other endpoints, which don't bother about `Accept` at all. If we add 406 to this endpoint, we should patch other ones too. IMO, let's just finally throw `warp` away and replace it with Axum first. (https://github.com/hyperledger/iroha/issues/3776)

There is another thing I should do now: some other endpoints return SCALE response, but they use (I guess, haven't checked) `Content-Type: application/octet-stream`. Should use the new MIME everywhere instead.

Reply:
Yeah, if `axum` will help with a general approach like this, it definitely makes sense to postpone this kind of changes

Reply:
I guess it should be split off to a different issue to track it?